### PR TITLE
Add FileSystemOptions and use in FileResolver

### PR DESCRIPTION
**Pull request by MarkArbogast**
Addresses: #2550 

Signed-off-by: Mark Arbogast <mark.t.arbogast@gmail.com>

### CODE REVIEW

File: src/main/java/io/vertx/core/VertxOptions.java [hunk 1]
```diff
@@ -705,7 +717,7 @@ public VertxOptions setAddressResolverOptions(AddressResolverOptions addressReso
    * @return wether the file resolver uses caching
    */
   public boolean isFileResolverCachingEnabled() {
-    return fileResolverCachingEnabled;
+    return fileSystemOptions.isFileResolverCachingEnabled();
```
Comment (vietj) [inline src/main/java/io/vertx/core/VertxOptions.java]: missing null check

File: src/main/java/io/vertx/core/VertxOptions.java [hunk 2]
```diff
@@ -715,7 +727,7 @@ public boolean isFileResolverCachingEnabled() {
    * @return a reference to this, so the API can be used fluently
    */
   public VertxOptions setFileResolverCachingEnabled(boolean fileResolverCachingEnabled) {
-    this.fileResolverCachingEnabled = fileResolverCachingEnabled;
+    this.fileSystemOptions.setFileResolverCachingEnabled(fileResolverCachingEnabled);
```
Comment (vietj) [inline src/main/java/io/vertx/core/VertxOptions.java]: missing null check

**Comment by vietj** [comment]
looks good, it should care of null file system options on the vertx options

**Comment by vietj** [comment]
we also want to deprecate the existing options on VertxOptions and say to use the nested file system options.

**Comment by MarkArbogast** [comment]
Added null checks and a unit test to verify them. Also deprecated `isFileResolverCachingEnabled` and `setFileResolverCachingEnabled`.

**Comment by pmlopes** [comment]
Question: should we also drop the system properties too? Does it make sense to allow changes at runtime? What takes precedence options or system properties?

**Comment by vietj** [comment]
@pmlopes that would be a breaking change, system properties should configure the default property values

File: src/main/java/io/vertx/core/file/FileSystemOptions.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
+
+/*
+ * Copyright (c) 2011-2017 Contributors to the Eclipse Foundation
```
Comment (vietj) [inline src/main/java/io/vertx/core/file/FileSystemOptions.java]: 2018

File: src/main/java/io/vertx/core/file/impl/FileResolver.java [hunk 4]
```diff
@@ -50,34 +51,38 @@
...
-  private static final boolean ENABLE_CP_RESOLVING = !Boolean.getBoolean(DISABLE_CP_RESOLVING_PROP_NAME);
   private static final String CACHE_DIR_BASE = System.getProperty(CACHE_DIR_BASE_PROP_NAME, DEFAULT_CACHE_DIR_BASE);
   private static final String JAR_URL_SEP = "!/";
   private static final Pattern JAR_URL_SEP_PATTERN = Pattern.compile(JAR_URL_SEP);
 
   private final File cwd;
   private File cacheDir;
   private Thread shutdownHook;
-  private final boolean enableCaching;
+  private final FileSystemOptions fileSystemOptions;
```
Comment (vietj) [inline src/main/java/io/vertx/core/file/impl/FileResolver.java]: keep the `enableCaching` field and instead assign its value from the provided options

File: src/main/java/io/vertx/core/file/impl/FileResolver.java [hunk 5]
```diff
@@ -104,7 +109,7 @@ public File resolveFile(String fileName) {
     if (cwd != null && !file.isAbsolute()) {
       file = new File(cwd, fileName);
     }
-    if (!ENABLE_CP_RESOLVING) {
+    if (!fileSystemOptions.isClassPathResolvingEnabled()) {
```
Comment (vietj) [inline src/main/java/io/vertx/core/file/impl/FileResolver.java]: add a final `enableCpResolving` field to the class instead

**Comment by vietj** [comment]
I've added a couple of extra comments, I think it needs to be also documented

**Comment by MarkArbogast** [comment]
Switched to use boolean variables instead of options in `FileResolver`. Updated documentation. Updated 2017 to 2018 in `FileSystemOptions` copyright comment.

**Comment by MarkArbogast** [comment]
Updated 2017 to 2018 in `FileSystemOptions` copyright comment.

**Comment by MarkArbogast** [comment]
Hi @vietj, anything else to fix for this PR?

File: src/test/java/io/vertx/test/core/VertxOptionsTest.java [hunk 6]
```diff
@@ -416,4 +424,15 @@ public void testJsonOptions() {
...
   }
+
+  @Test
+  public void testNullFileSystemOptions() {
+    VertxOptions options = new VertxOptions().setFileSystemOptions(null);
+    try {
+      options.isFileResolverCachingEnabled();
+      options.setFileResolverCachingEnabled(true);
+    } catch (NullPointerException e) {
+      fail("Should handle null FileSystemOptions");
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/VertxOptionsTest.java]: just let it go through and let the NPE fail the test

**Comment by MarkArbogast** [comment]
Updated `testNullFileSystemOptions` to fail directly instead of catching NPE.

**Comment by vietj** [comment]
thanks for the contrib!

**Pull request merged**